docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,674 @@
+Again, the easiest way to do this is to `hlc.Update(txn.ReadTimestamp)` before
+allowing a refreshed/retried transaction to read at its new timestamp. This
+again becomes a no-op when the new ReadTimestamp is "real" and not "synthetic".
+
+### Future-Time Closed Timestamps
+
+To this point, we've talked about writing in the future as a means to reduce
+contention, but doing so has a second major benefit – it enables non-stale
+follower reads if coupled with a closed timestamp mechanism that closes off
+timestamps in the future.
+
+This RFC proposes that "non-blocking Ranges" use a separate closed timestamp
+tracker that is configured to close time out at least WAN RTT + max_offset in
+the future. This ensures that all followers in these ranges will hear about and
+active closed timestamps above "present time" so that they can serve all
+"present time" reads locally.
+
+The details of this change are TBD.
+
+#### Aside: Writing in Future vs. Closing Time in Future
+
+It may be evident to readers that there is some inherent but fuzzy connection
+between the idea of writing in the future and closing time in the future. It is
+interesting to point out that in systems like Spanner and YB which assign
+monotonically increasing timestamps to all writes within a replication group and
+therefore "instantaneously" close out time, these two operations are one in the
+same.
+
+CockroachDB's transaction and replication model are not quite set up to support
+this. Transactions in CockroachDB are optimized for committing at their original
+read timestamp, so they don't like when their writes get bumped. This allows for
+optimistic, lockless reads. Replication in CockroachDB is below the level of
+evaluation (see [proposer evaluated kv](20160420_proposer_evaluated_kv.md)), so
+the timestamp of a replicated value need to be determined before
+replication-level sequencing. This allows for evaluation-level parallelism and
+less need for determinism during evaluation.
+
+For these reasons, the closed timestamp of a Range lags the timestamps of writes
+in a Range and the closed timestamp is more of a Range-level concern and less of
+a Request-level concern in CockroachDB (i.e. a single request cannot dictate the
+closed timestamp for a Range). This is why the closed timestamp dictates the
+time at which requests perform writes (as we'll see in the next section) instead
+of the other way around.
+
+### Becoming a Non-Blocking Transaction
+
+As stated earlier, this proposal does not expose "non-blocking transactions" to
+users directly. Instead, it exposes "non-blocking Ranges" to users. It was also
+mentioned earlier that a "non-blocking Range" will have a closed timestamp
+tracker that is leading "present time" by some duration. As it turns out,
+configuring this closed timestamp tracker to lead "present time" by
+`non_blocking_duration` is enough to implement "non-blocking transactions"
+without any other server-side changes.
```

Comment:
There are still changes here, no? We don't do per-range closed timestamps at the moment, so we need the infra for that. But it'd be nice if those were the only changes. I think the quiescence stuff will need some work as it's tied to the store-wide closed timestamp.